Syntactic sugar, typedefs as components, and timesteps

player-03 opened the pull request:
I've chosen to build a game engine using Echoes, which required some additions to the base library.

Abstract entities
---------------------
This is syntactic sugar, designed to help ease object-oriented programmers into the ECS mindset. Using an abstract that wraps `Entity`, you can make it act more like a class:

```haxe
abstract Soldier(Entity) {
    public var position:Point;
    public var sprite:Image;
    public var offense:Offense;
    public var defense:Defense;
    public var hp:HP;
    
    public function new(x:Float, y:Float) {
        this = new Entity();
        position = new Point(x, y);
        sprite = new Image("soldier.png");
        offense = 4;
        defense = 2;
        hp = new HP(10);
    }
}
```

It looks and feels like a class, but it's still an `Entity`. For instance, say the soldier's armor breaks and you update the variables. Macro magic will turn your code into the corresponding `Entity` functions:

```haxe
soldier.defense = null; //soldier.remove(Defense);
soldier.hp.value -= 2; //soldier.get(HP).value -= 2;
soldier.sprite = new Image("soldier_broken_armor.png"); //solder.add(new Image("soldier_broken_armor.png"));
```

Other features:

- Every `Entity` function is forwarded, so you can call `add()` and `remove()` and so on.
- You can "extend" your own abstracts to create more specialized entities. For instance, you could have `abstract Sniper(Soldier)`, `abstract Cavalry(Soldier)`, and `abstract Defender(Soldier)`, each with a different set of components. (The basic `Soldier` components will remain available to all three.)
- You can write getters and setters for all your components. (Though since I have no use for this feature, it's been a while since I tested it.)
- You can write other functions, as normal for an `abstract` type. Most code should of course go in systems, but short convenience functions still have their place.

Typedefs as components
---------------------------------
Consider the `Offense` and `Defense` components above. They take integer values, which can be done using abstracts:

```haxe
abstract Offense(Int) from Int to Int {}
abstract Defense(Int) from Int to Int {}
```

But that feels like too much typing for something so simple. Wouldn't it be nice if Haxe had a built-in feature for this sort of situation? Something easier to read and [less repetitive](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)?

```haxe
typedef Offense = Int;
typedef Defense = Int;
```

So now Echoes recognizes these as individual components.

Timesteps
--------------
Timekeeping is fundamental to games, and kind of complicated. Most code just needs to be run once per frame, but some code needs to be multiple times, based on the number of milliseconds that have passed since last frame. Some customization is needed.

Safeguards are needed too. You usually want a maximum time per frame, in case the operating system goes to sleep. Otherwise, the game might wake up after an hour and try to calculate a full hour's worth of physics.

(I know this feels like feature creep. Echoes is supposed to be simple and lightweight. But the fact of the matter is, as long as Echoes has `@update` functions, it's responsible for this. The only alternative is to fundamentally change - or allow overriding of - `SystemList.__update__()`, which is a terrible idea.)

The `Timestep` class is the least invasive solution I could think of. For more details, see the class itself.

player-03 committed:
Allow using typedefs. (Requires explicit typing.)

If you call Entity.add(x) without explicitly typing x, the compiler
will follow the typedef and use the underlying type instead.
To avoid this, either declare your variables using a colon
(e.g. "var x:MyTypedefType;"), or cast them when you call add()
(e.g. "entity.add((x:MyTypedefType));").

player-03 committed:
Implement AbstractEntity.

player-03 committed:
Condense AbstractEntity.

player-03 committed:
Fix instance variables in abstract entities

player-03 committed:
Fix missing removeIfMatched() function.

player-03 committed:
Implement Timestep.

This allows a nested SystemList to use a different timestep than its
parent. For instance, the parent could use the full length of the frame,
while the child could have a fixed timestep.

player-03 committed:
Abstract entity bug fixes.

When setting a value to null, actually remove it from the entity.

Improve the error message that appears when using an unsupported type.

Fix baseTypeToString(), which had previously been untested.

Make sure Null<Type> is always converted to Type, instead of sometimes.

player-03 committed:
Account for null values on static platforms.

player-03 committed:
Avoid generating excessively long class names.

Some targets create intermediate files for each generated class, and
most operating systems set a limit on filename length.

TODO: use Compiler.getDefine() to determine the target, and only run
this code if necessary. (#if doesn't work because this is a macro.)

player-03 committed:
Use the decorator pattern for `Timestep`.

Plus, add documentation. I know most of the classes don't come with any,
but you have to start somewhere.

player-03 committed:
Implement `ScaledTimestep`.

This is why the decorator pattern is necessary.

player-03 committed:
Allow pausing `Timestep`s.

player-03 committed:
Allow components with type parameters.

Due to macro limitations, special syntax is required for retrieval:
`entity.get((_:Array<String>))`

player-03 committed:
Add type parameter support to `AbstractEntity`.

player-03 committed:
Prepare `parseComplexType()` for merging.

player-03 committed:
Move entity macro functions to `EntityTools`.

Haxe 4 doesn't allow macros to call macros, so these four functions
needed to be made static instead.

player-03 committed:
Merge branch 'abstractentity' into dev

player-03 committed:
Handle `entity.add(new MyTypedef())`.

Without an explicit workaround, `Context.typeof()` will follow the
typedef, losing that extra information.

`MacroTools` had a function for this, but it's (seemingly) only needed
in one place, so I removed the function to save space.

player-03 committed:
Strip trailing whitespace.

deepcake commented:
Hello!

Unfortunately, I no longer have any motivation or time to spend on this framework. And I would rather leave it as it is, with backward compatibility. 

You can still freely use any code from this framework in your projects.

Sorry if I ruined your expectations or something.

Good luck with your game engine!

player-03 commented:
The typedefs are the only things here that break backwards compatibility.

Perhaps we could make it opt-in: you have to tag them with `@:typedefComponent` to get the new behavior, but by default it'll behave as it always did. (And then I'll use [`addGlobalMetadata()`](https://api.haxe.org//haxe/macro/Compiler.html#addGlobalMetadata) to modify the default behavior for my own projects.)

player-03 committed:
Optimize `entity.add()`.

Instead of attempting to add an entity to ALL views, it's best to
check only views of the new component(s).

player-03 committed:
Strip trailing whitespace.

player-03 commented:
Hi again @deepcake!

I respect your decision to leave this as the final version, so I'll withdraw this pull request. I plan to continue working on [my fork](https://github.com/player-03/echoes), just without the expectation that you'll eventually merge it.

My question is, do you mind if I continue using the name "Echoes"? I really do like the name, but I'll respect your decision either way. Corollary: is it ok to upload a Haxelib with this name? (You'd be a co-author, with permission to release updates.) It's just important to me that this eventually makes its way to Haxelib.

deepcake commented:
Hi!

Yeah, you can freely use the name and upload your version to Haxelib.

player-03 commented:
Wonderful, thanks!
